File: backend/app/routes/transactions.py
"""
Transaction API Routes.
"""

from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query, status
from app.schemas.transactions import (
    TransactionCreate,
    TransactionResponse,
    TransactionSummaryResponse,
    TransactionType
)
from app.services.transaction_service import transaction_repository

router = APIRouter(prefix="/transactions", tags=["Transactions"])


@router.post(
    "",
    response_model=TransactionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Record a new transaction"
)
def create_transaction(payload: TransactionCreate):
    """
    Add a new income or expense transaction.
    Accepts category, amount, date, recurring settings, and notes.
    """
    return transaction_repository.create(payload)


@router.get(
    "",
    response_model=List[TransactionResponse],
    summary="List all recorded transactions"
)
def list_transactions(
    transaction_type: Optional[TransactionType] = Query(None, description="Filter by 'income' or 'expense'"),
    category: Optional[str] = Query(None, description="Filter by category name"),
    is_recurring: Optional[bool] = Query(None, description="Filter by recurring flag")
):
    """Retrieve all recorded transactions with optional filtering."""
    return transaction_repository.get_all(
        transaction_type=transaction_type,
        category=category,
        is_recurring=is_recurring
    )


@router.get(
    "/summary",
    response_model=TransactionSummaryResponse,
    summary="Get aggregated transaction summary"
)
def get_transaction_summary():
    """
    Calculate and return:
    - Total income
    - Total expenses
    - Net savings & savings rate (%)
    - Monthly estimated cash flows
    - Category-wise spending breakdown
    - Recurring commitments breakdown
    """
    return transaction_repository.get_summary()


@router.get(
    "/{txn_id}",
    response_model=TransactionResponse,
    summary="Get a transaction by ID"
)
def get_transaction(txn_id: str):
    record = transaction_repository.get_by_id(txn_id)
    if not record:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Transaction with ID '{txn_id}' not found."
        )
    return record


@router.delete(
    "/{txn_id}",
    status_code=status.HTTP_200_OK,
    summary="Delete a transaction"
)
def delete_transaction(txn_id: str):
    success = transaction_repository.delete(txn_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Transaction with ID '{txn_id}' not found."
        )
    return {"success": True, "message": f"Transaction {txn_id} successfully deleted."}

File: frontend-old-backup/src/lib/api/transactions.ts
import { apiRequest, shouldUseMockData } from "./client";

export interface TransactionPayload {
  title: string;
  amount: number;
  category: string;
  transaction_type: "income" | "expense";
  date?: string;
  is_recurring?: boolean;
  notes?: string;
}

export async function getTransactions(params?: {
  transaction_type?: string;
  category?: string;
  is_recurring?: boolean;
}) {
  if (shouldUseMockData()) {
    return [];
  }
  const query = new URLSearchParams();
  if (params?.transaction_type) query.append("transaction_type", params.transaction_type);
  if (params?.category) query.append("category", params.category);
  if (params?.is_recurring !== undefined) query.append("is_recurring", String(params.is_recurring));

  const qs = query.toString();
  return apiRequest<any[]>(`/api/v1/transactions${qs ? '?' + qs : ''}`);
}

export async function createTransaction(payload: TransactionPayload) {
  if (shouldUseMockData()) {
    return { id: "txn-" + Date.now(), ...payload };
  }
  return apiRequest<any>("/api/v1/transactions", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getTransactionSummary() {
  if (shouldUseMockData()) {
    return {
      total_income: 85000,
      total_expenses: 35000,
      net_savings: 50000,
      savings_rate_pct: 58.8,
      transaction_count: 0,
    };
  }
  return apiRequest<any>("/api/v1/transactions/summary");
}
